vulyk/blueprints/gamification/models/state.py

```diff
@@ -80,6 +80,54 @@ def from_state(cls, state: UserState):
             last_changed=state.last_changed
         )
 
+    @classmethod
+    def get_or_create_by_user(cls, user: User) -> UserState:
+        """
+        Returns an existing UserStateModel instance bound to the user, or a new
+        one if didn't exist before.
+
+        :param user: Current User model instance
+        :type user: User
+
+        :return: UserState instance
+        :rtype: UserState
+        """
+        try:
+            state_model = cls.objects.get(user=user)
+        except cls.DoesNotExist:
+            state_model = cls.objects.create(user=user)
+
+        return state_model.to_state()
+
+    @classmethod
+    def update_state(cls, diff: UserState) -> None:
+        """
+        Prepares and conducts an atomic update query from passed diff.
+
+        :param diff: State object contains values that are to be changed only.
+        :type diff: UserState
+
+        :rtype: None
+        """
+        update_dict = {'set__last_changed': diff.last_changed}
+
+        if diff.level > 0:
+            update_dict['set__level'] = diff.level
+        if diff.points > 0:
+            # DON'T PUT YOUR BLAME ON ME
+            # https://github.com/MongoEngine/mongoengine/blob/master/mongoengine/fields.py#L415
+            update_dict['inc__points'] = float(diff.points)
+        if diff.actual_coins != 0:
+            # ALL YOUR DECIMALS ARE BELONG TO US
+            update_dict['inc__actual_coins'] = float(diff.actual_coins)
```

Comment:
I have found [this](https://github.com/MongoEngine/mongoengine/blob/master/tests/queryset/queryset.py#L1877) and extrapolated. The first test on donation will figure it out.